magpie/adapter/__init__.py

```diff
@@ -243,3 +257,83 @@ def configurator_factory(self, container):  # noqa: R0201
+        for hook_cfg in svc_hooks:
+            if hook_cfg["type"] != hook_type:
+                continue
+            if hook_cfg["method"] not in ["*", method]:
+                continue
+            hook_path = normalize_field_pattern(hook_cfg["path"])
+            if not re.match(hook_path, path):
+                continue
+            hook_query = normalize_field_pattern(hook_cfg["query"])
+            if not re.match(hook_query, query):
+                continue
+            hook_target = import_target(hook_cfg["target"])
+            hook_qs = "?" + query if query else ""
+            if not hook_target:
+                LOGGER.warning("Hook matched %s (%s %s%s) but specified target [%s] could not be loaded.",
+                               hook_type, method, path, hook_qs, hook_cfg["target"])
+                continue
+            LOGGER.debug("Hook matched %s (%s %s%s) [%s]", hook_type, method, path, hook_qs, hook_cfg["target"])
+            signature = inspect.Signature(hook_target)
+            kwargs = {}
+            if len(signature.parameters) > 1:
+                hook = copy.deepcopy(hook_cfg)
+                for key, val in [("service", svc_config), ("hook", hook)]:
+                    if key in signature.parameters:
+                        kwargs[key] = val
+            try:
+                instance = hook_target(instance, **kwargs)
+            except Exception as exc:
+                LOGGER.error("Hook failed %s (%s %s%s) [%s]",
+                             hook_type, method, path, hook_qs, hook_cfg["target"], exc_info=exc)
+                raise exc
+        return instance
+
+    def request_hook(self, request, service):
+        # type: (Request, ServiceConfig) -> Request
+        """
+        Apply modifications onto the request before sending it.
+
+        .. versionadded:: 3.25
+            Requires ``Twitcher >= 0.7.x``.
+
+        Request members employed after this hook is called include:
+        - :meth:`Request.headers`
+        - :meth:`Request.method`
+        - :meth:`Request.body`
+
+        This method can modified those members to adapt the request for specific service logic.
+        """
+        request = self._apply_hooks(
+            request, service["name"], "request",
+            request.method, request.path, request.query_string
+        )
+        return request
+
+    def response_hook(self, response, service):
+        # type: (Response, ServiceConfig) -> Response
+        """
+        Apply modifications onto the response from sent request.
+
+        .. versionadded:: 3.25
+            Requires ``Twitcher >= 0.7.x``.
+
+        The received response from the proxied service is normally returned directly.
+        This method can modify the response to adapt it for specific service logic.
+        """
+        response = self._apply_hooks(
+            response, service["name"], "response",
+            response.request.path, response.request.path, response.request.query_string
```

Comment:
Should the first argument of this line be `response.request.method`?

Reply:
Indeed, was fixed in commit not yet pushed.